hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,611 @@
+# Meta
+[meta]: #meta
+- Name: Hera Decorator Syntax
+- Start Date: 2024-04-13
+- Author(s): [@samj1912](https://github.com/elliotgunton), [@elliotgunton](https://github.com/elliotgunton)
+- Supersedes: (put "N/A" unless this replaces an existing HEP, then link to that HEP)
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Code Examples](#code-examples)
+  - [Basic Workflow](#basic-workflow)
+  - [DAG Template](#dag-template)
+  - [Steps Template](#steps-template)
+  - [Container Template](#container-template)
+  - [Workflow Template Ref](#workflow-template-ref)
+  - [How to teach (OPTIONAL)](#how-to-teach-optional)
+- [Implementation (OPTIONAL)](#implementation-optional)
+  - [Link to the Implementation PR](#link-to-the-implementation-pr)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions (OPTIONAL)](#unresolved-questions-optional)
+
+# Overview
+[overview]: #overview
+
+This HEP introduces a new and powerful decorator based syntax for defining all kinds of Workflows and Templates.
+
+# Motivation
+[motivation]: #motivation
+
+The current context manager based syntax is good (especially for bridging the gap from Python to Argo's YAML) but it doesn't map well to template types that contain inputs/outputs. With the `@script` decorator, we made Hera feel extremely Pythonic and it has the added benefit of being runnable locally. We wish to extend this syntax to other template types, namely DAGs, Steps and Containers, which will allow for better readability and local testing. In this HEP, we also show how we will be able to generate stubs for WorkflowTemplates and improvements for referencing TemplateRefs under the new decorator syntax.
+
+# Proposal
+
+This HEP proposes that Argo's DAG, Steps and Container templates be mapped into Hera as decorated functions with HeraIO classes as the inputs and outputs. We will also introduce a new script decorator under the `hera.workflows.workflow.Workflow` class which will enforce use of the script runner along with the HeraIO classes.
+
+> ⚠️ The existing context manager syntax, and the script decorator, will continue to be supported, but they are unlikely
+> to receive any active development of new features.
+
+# Code Examples
+
+## Basic Workflow
+
+```python
+from hera.workflows import WorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# Users must subclass `hio.Input` to define the template's inputs
+class MyInput(hio.Input):
+    user: str
+
+@wt.entrypoint  # Sets hello_world as the default entrypoint for the workflow template, it is an error to set entrypoint on multiple functions
```

Comment:
`entrypoint` is already a property of `Workflow` - do you see a sensible way to refactor and allow this decorator name or should we just use `@wt.set_entrypoint`?

Reply:
It should be possible to make entrypoint a setter and then have dynamic behavior based on the input type.